plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintFormatTask.kt

```diff
@@ -1,13 +1,31 @@
 package org.jlleitschuh.gradle.ktlint
 
+import org.gradle.api.file.FileTree
 import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.OutputFiles
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.SkipWhenEmpty
 import org.gradle.process.JavaExecSpec
 import javax.inject.Inject
 
+@CacheableTask
 open class KtlintFormatTask @Inject constructor(
     objectFactory: ObjectFactory
 ) : KtlintCheckTask(objectFactory) {
     override fun additionalConfig(): (JavaExecSpec) -> Unit = {
         it.args("-F")
     }
+
+    /**
+     * Fixes the issue when input file is restored to pre-format state and running format task again fails
+     * with "up-to-date" task state.
+     *
+     * Note this approach sets task to "up-to-date" only on 3rd run when both input and output sources are the same.
+     */
```

Comment:
@nedtwigg I know you fixed this same problem with Spotless. Care to offer any insights into how you resolved it?

Reply:
I wrote it up [here](https://discuss.gradle.org/t/two-cases-where-the-incremental-api-is-lacking/24266), and included links to the PR's that fixed it.

Reply:
It is kind of complex solution, I'd prefer not introduce it here.